### PR TITLE
Update leveldb to bitcoin's upstream.

### PR DESCRIPTION
**Pull request by sinetek**
This issue is linked with #2531.

Would this work to solve the issue?

### CODE REVIEW

**Comment by rnicoll** [comment]
I've just realised this is raised against `master` - can you rebase on `1.14.5-dev` please and then edit the PR to point to that branch as destination? Thanks

**Comment by patricklodder** [comment]
> can you rebase on 1.14.5-dev please and then edit the PR to point to that branch as destination? 

I didn't change the base myself because I'm not sure if this can make it into 1.14.5. 

Would be a waste to do it twice? I'm still testing scenarios right now

**Comment by sinetek** [comment]
> I've just realised this is raised against `master` - can you rebase on `1.14.5-dev` please and then edit the PR to point to that branch as destination? Thanks

Absolutely, I can do all the work that's needed for this, point me to the branch you'd like.

**Comment by patricklodder** [comment]
I've done a number of tests with this trying to get a feel for portability cross-version when we do this. I have not been able to trigger any crash or force a reindex when switching back-and-forth between releases. Will try a couple more things between this PR, 1.14.4 and 1.14.5 but so far this looks pretty good!

**Comment by sinetek** [comment]
> I've done a number of tests with this trying to get a feel for portability cross-version when we do this. I have not been able to trigger any crash or force a reindex when switching back-and-forth between releases. Will try a couple more things between this PR, 1.14.4 and 1.14.5 but so far this looks pretty good!

Great, thanks!

**Comment by rnicoll** [comment]
I'm a bit concerned this is significantly beyond the version Bitcoin Core uses. It looks like it's `master` from LevelDB, whereas the latest release is 1.23, and it sounds like Bitcoin Core uses 1.22. I'd definitely want to stick to released versions we know are rock solid, not use the cutting edge.

If there's no overriding reason, can we move to 1.22 first and make sure that works well, before we go any further?

**Comment by patricklodder** [comment]
> If there's no overriding reason, can we move to 1.22 first and make sure that works well, before we go any further?

@rnicoll ed0bc9c is 20a6babf, which is also part of 1.21-dev? Which commit do you see LevelDB master in?

**Comment by rnicoll** [comment]
@patricklodder I was comparing files on disk, but wonder if I had the wrong branch or something, because what I see in GitHub doesn't match what I was seeing. Hangon trying that again.

**Comment by rnicoll** [comment]
@patricklodder Ah, I see; yes, it matches Bitcoin, I was testing against LevelDB's releases directly, and they're not a single specific LevelDB version, more of a mix.

**Comment by patricklodder** [comment]
> they're not a single specific LevelDB version, more of a mix.

That makes sense - I am quite sure bitcoin/bitcoin merges from https://github.com/bitcoin-core/leveldb-subtree that they maintain rather than upstream levelDB

**Comment by rnicoll** [comment]
With the conflicts resolved as per https://github.com/rnicoll/dogecoin/tree/newldb, I've done:

* Load an existing database for testnet and leave it to sync
* Reindex the fully synced database
* Load the reindexed database into an old version just to see if I can get anything to break

I think that's good to merge after rebase, and then put onto long running nodes to soak test just in case. What do you think @patricklodder ?

**Comment by patricklodder** [comment]
> I think that's good to merge after rebase, and then put onto long running nodes to soak test just in case.

I've done those same testing strategies. One thing I thought we should do: 1.21 has a leveldb migration process - we should test that. Your work branch is much further than what I have though - are you able to check if that still works with this version as a source?

**Comment by sinetek** [comment]
> > they're not a single specific LevelDB version, more of a mix.
> 
> That makes sense - I am quite sure bitcoin/bitcoin merges from https://github.com/bitcoin-core/leveldb-subtree that they maintain rather than upstream levelDB

There doesn't seem to be any upstream changes between 0.20 // 0.21 // 0.22 // master in what is the leveldb tree.

**Comment by rnicoll** [comment]
> 
> 
> > I think that's good to merge after rebase, and then put onto long running nodes to soak test just in case.
> 
> I've done those same testing strategies. One thing I thought we should do: 1.21 has a leveldb migration process - we should test that. Your work branch is much further than what I have though - are you able to check if that still works with this version as a source?

Yup, did a sync with this version and upgraded to 1.21, worked fine

**Comment by patricklodder** [comment]
@sinetek could you please rebase against 1.14.6-dev to (hopefully auto) reapply the configure.ac patch from upstream? Then I think we're good to merge. It'll see a lot of runs in the CI once merged as we're early in the release cycle right now.

**Comment by sinetek** [comment]
> @sinetek could you please rebase against 1.14.6-dev to (hopefully auto) reapply the configure.ac patch from upstream? Then I think we're good to merge. It'll see a lot of runs in the CI once merged as we're early in the release cycle right now.

Yes I will do it, just a moment please.

**Comment by patricklodder** [comment]
> There doesn't seem to be any upstream changes, from bitcoin/bitcoin that is, between 0.20 // 0.21 // 0.22 // master in what is the leveldb tree.

To make 100% sure, I did a compare between https://github.com/bitcoin-core/leveldb-subtree and src/leveldb from this PR.

script:
```bash
git checkout https://github.com/bitcoin-core/leveldb-subtree.git
git checkout https://github.com/dogecoin/dogecoin.git
pushd dogecoin
gh pr checkout 2637
diff -dur ../leveldb-subtree src/leveldb
popd
```

diff output:
```
Only in ../leveldb-subtree: .git
diff -dur ../leveldb-subtree/db/version_set.h src/leveldb/db/version_set.h
--- ../leveldb-subtree/db/version_set.h	2021-11-25 19:09:30.330778758 +0000
+++ src/leveldb/db/version_set.h	2021-11-22 02:00:30.647043178 +0000
@@ -371,7 +371,7 @@
   // Each compaction reads inputs from "level_" and "level_+1"
   std::vector<FileMetaData*> inputs_[2];  // The two sets of inputs
 
-  // State used to check for number of overlapping grandparent files
+  // State used to check for number of of overlapping grandparent files
   // (parent == level_ + 1, grandparent == level_ + 2)
   std::vector<FileMetaData*> grandparents_;
   size_t grandparent_index_;  // Index in grandparent_starts_
```

I think I can live with this diff 😁

File: configure.ac
Comment (patricklodder) [inline configure.ac:390]: those lines should be removed, we don't have this target right now. Could add it in another pull, but I think we'd need more work for `*-linux-android` to work?

**Review by patricklodder — CHANGES_REQUESTED** [review]
One minor comment on the last commit, and needs a rebase to resolve conflict with `configure.ac`, otherwise tested and good to go.

File: configure.ac
Comment (sinetek) [reply configure.ac:390]: OK I will remove it here.

**Comment by sinetek** [comment]
> One minor comment on the last commit, and needs a rebase to resolve conflict with `configure.ac`, otherwise tested and good to go.

Done and done.

**Review by patricklodder — APPROVED** [review]
tested ACK

- Code review ✅
- [diff with Bitcoin's subtree](https://github.com/dogecoin/dogecoin/pull/2637#issuecomment-979432120) ✅
- Tested upgrade from v1.14.5 ✅
- Tested upgrade to current 1.21-dev ✅

**Pull request merged**
